File: frontend/src/lib/utils/admin-auth.js
// Утилиты для авторизации администраторов

const API_BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000';

export const adminAuth = {
  /**
   * Вход администратора
   */
  async login(username, password) {
    try {
      const response = await fetch(`${API_BASE_URL}/api/admin/login/`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ username, password }),
      });

      if (!response.ok) {
        const error = await response.json();
        throw new Error(error.error || 'Ошибка авторизации');
      }

      const data = await response.json();
      
      // Сохраняем токен и данные пользователя
      localStorage.setItem('admin_token', data.token);
      localStorage.setItem('admin_user', JSON.stringify(data.user));
      
      return data;
    } catch (error) {
      console.error('Login error:', error);
      throw error;
    }
  },

  /**
   * Проверка токена
   */
  async verify() {
    const token = localStorage.getItem('admin_token');
    
    if (!token) {
      return false;
    }

    try {
      const response = await fetch(`${API_BASE_URL}/api/admin/verify/`, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
      });

      if (!response.ok) {
        // Токен невалиден - очищаем
        this.logout();
        return false;
      }

      return true;
    } catch (error) {
      console.error('Verify error:', error);
      this.logout();
      return false;
    }
  },

  /**
   * Выход
   */
  logout() {
    localStorage.removeItem('admin_token');
    localStorage.removeItem('admin_user');
  },

  /**
   * Получить текущего пользователя
   */
  getUser() {
    const userStr = localStorage.getItem('admin_user');
    return userStr ? JSON.parse(userStr) : null;
  },

  /**
   * Получить токен
   */
  getToken() {
    return localStorage.getItem('admin_token');
  },

  /**
   * Проверка авторизации (синхронно)
   */
  isAuthenticated() {
    return !!this.getToken();
  }
};

File: backend/gooddrive_backend/admin_auth.py
"""
Модуль для авторизации администраторов
"""
from rest_framework.decorators import api_view, permission_classes
from rest_framework.permissions import AllowAny
from rest_framework.response import Response
from rest_framework import status
from django.contrib.auth import authenticate
from django.contrib.auth.models import User
from datetime import datetime, timedelta
import jwt
from django.conf import settings


# Секретный ключ для JWT (в продакшене использовать из .env)
JWT_SECRET = getattr(settings, 'SECRET_KEY', 'your-secret-key-change-in-production')
JWT_ALGORITHM = 'HS256'
JWT_EXPIRATION_HOURS = 24


@api_view(['POST'])
@permission_classes([AllowAny])
def admin_login(request):
    """
    Авторизация администратора
    
    POST /api/admin/login/
    {
        "username": "admin",
        "password": "password"
    }
    """
    username = request.data.get('username')
    password = request.data.get('password')
    
    if not username or not password:
        return Response(
            {'error': 'Введите логин и пароль'},
            status=status.HTTP_400_BAD_REQUEST
        )
    
    # Аутентификация через Django
    user = authenticate(username=username, password=password)
    
    if user is None:
        return Response(
            {'error': 'Неверный логин или пароль'},
            status=status.HTTP_401_UNAUTHORIZED
        )
    
    # Проверяем, что пользователь - администратор
    if not user.is_staff:
        return Response(
            {'error': 'Доступ запрещён. Требуются права администратора'},
            status=status.HTTP_403_FORBIDDEN
        )
    
    # Генерируем JWT токен
    payload = {
        'user_id': user.id,
        'username': user.username,
        'email': user.email,
        'is_superuser': user.is_superuser,
        'exp': datetime.utcnow() + timedelta(hours=JWT_EXPIRATION_HOURS),
        'iat': datetime.utcnow()
    }
    
    token = jwt.encode(payload, JWT_SECRET, algorithm=JWT_ALGORITHM)
    
    return Response({
        'token': token,
        'user': {
            'id': user.id,
            'username': user.username,
            'email': user.email,
            'first_name': user.first_name,
            'last_name': user.last_name,
            'is_superuser': user.is_superuser
        }
    })


@api_view(['POST'])
def admin_verify(request):
    """
    Проверка токена администратора
    
    POST /api/admin/verify/
    Headers: Authorization: Bearer <token>
    """
    auth_header = request.headers.get('Authorization', '')
    
    if not auth_header.startswith('Bearer '):
        return Response(
            {'error': 'Токен не предоставлен'},
            status=status.HTTP_401_UNAUTHORIZED
        )
    
    token = auth_header.split(' ')[1]
    
    try:
        payload = jwt.decode(token, JWT_SECRET, algorithms=[JWT_ALGORITHM])
        
        # Проверяем существование пользователя
        try:
            user = User.objects.get(id=payload['user_id'])
            if not user.is_staff:
                return Response(
                    {'error': 'Доступ запрещён'},
                    status=status.HTTP_403_FORBIDDEN
                )
            
            return Response({
                'valid': True,
                'user': {
                    'id': user.id,
                    'username': user.username,
                    'email': user.email,
                    'is_superuser': user.is_superuser
                }
            })
        except User.DoesNotExist:
            return Response(
                {'error': 'Пользователь не найден'},
                status=status.HTTP_401_UNAUTHORIZED
            )
            
    except jwt.ExpiredSignatureError:
        return Response(
            {'error': 'Токен истёк'},
            status=status.HTTP_401_UNAUTHORIZED
        )
    except jwt.InvalidTokenError:
        return Response(
            {'error': 'Неверный токен'},
            status=status.HTTP_401_UNAUTHORIZED
        )


@api_view(['POST'])
def admin_logout(request):
    """
    Выход администратора (на клиенте удаляем токен)
    """
    return Response({'message': 'Выход выполнен успешно'})

